README.md
Alarm with Weather
=====

About
-----

Alarm is a small shell script that can aid you in alerting you of things at
specified times. It does this by displaying a message in X (using one of
the variety of X message displaying tools) at the time of the alert.

It supports a very flexible time input format (i.e. 'tomorrow', '5 may
2005', 'tomorrow 20:00'), one-time alerts and repeated alerts.


Installation and setup
----------------------

### Requirements

Alert requires the following to be installed:(X11설치)
		
*   X11
		
And at least ONE OF THE FOLLOWING programs to be installed: (이 프로그램들 중 적어도 하나는 설치해야합니다.)

*   xmessage 
*   dialog in combination with a terminal emulator 
*   gdialog


### Installation

1.  Copy the 'alarm' script to a directory that is present in the path.
2.  Add the following line to the .xinitrc file in your home directory:
		
    alarm --daemon&
	
Usage
-----

Alarm can run in two modes. The first is the daemon mode in which it will
notify you of events when they occur and the other is the edit mode through
which you can add, change, delete and list events. Alarm in daemon mode
should be run under X.

Below is the output of alarm --help:
	
	Usage: alarm [ACTION] [PARAMETERS]                          
                                                               
	Actions:
	
	  Additionally required parameters are listed in parenthesis.
	  
	  -n, --new    (-t, -m)       Create a new alert at timespec   
	  -c, --change (-i, -t, -m)   Change the alert identified by id
	  -d, --delete (-i)           Delete the alert identified by id
	  -l, --list   ([-i])         List alert(s)                    
		  --daemon            Start the alert daemon           
		  --weather           Show weather
                  --sun               Show sunrise and sunset														   
	Parameters                                                     
	  -t, --timespec TIMESPEC     Date/time specification at which 
								  the alert should be triggered    
	  -m, --message MESSAGE       Message that should be displayed 
	  -i, --id ID                 Alarm ID that should be changed, 
								  deleted or listed                
	  -f, --file FILE             Use FILE as alarm file instead of
								  the default ~/.alarms            
	  -r, --repeat                Repetative alarm                 

### Adding alerts

Here are some examples of adding one-time alerts:
	
	$ alarm -n -t "1 april 2005" -m "Watch out for any pranks." 
	$ alarm -n -t "20:15" -m "Appointment with Some Person" 
	$ alarm -n -t "+15 minutes" -m "Call back client" 
	$ alarm -n -t "tomorrow" -m "Do a good deed" 

Here are some examples of adding repeating alerts:
	
	$ alarm -n -r -t "friday" -m "Every Friday: beer-drinking competition"
	$ alarm -n -r -t "0:00" -m "Hey\! It's twelve o'clock. Get to sleep\!"

### Listing set alarms

Here is an example of the output of alerts:

	$ alarm -l

	 id | m |                 date | message
	----+---+----------------------+---------------------------------------------
	  1 |   |     2005-04-01 00:00 | Watch out for any pranks.
	  2 |   |     2005-04-11 20:15 | Appointment with Some Person
	  3 |   |     2005-04-11 16:31 | Call back client
	  4 |   |     2005-04-12 16:16 | Do a good deed
	  5 | r |               friday | Every Friday: beer-drinking competition
	  6 | r |                 0:00 | Hey! It's twelve o'clock. Get to sleep!

	
The fields are:

    id      = Notification ID (use with -i).
    m       = Mode ('n'ormal, 'r'ecurring).
    date    = The absolute ('2005-04-11 20:00') or relative ('friday') date & 
              time of the event.
    message = The message to show when the event is triggered.

### Changing alerts

Here is an example of how to change an alert:
	
	$ alarm -c -i2 -t "thursday" -m "Thursday beer-drinking competition tonight"

### Deleting alerts

Here is an example of deleting an alert.

	$ alarm -d -i2
	
### show sunrise, sunset
    
        $ alarm --sun
	
## show weather

        $ alarm --weather

Copyright
---------

Copyright (C), 2005-2013 Ferry Boender. Released under the General Public
License For more information, see the COPYING file supplied with this program.                                                          



ChangeLog
ChangeLog for alarm.sh
v0.3 [INUMONG] [Sat Dec 7 18:24:30 CET 2019]
        * add sunrise, sunset, weather

v0.2 [FB] [Sun Jan  8 23:53:56 CET 2006]
	* No output shown when in daemon mode anymore.
	* Alarm.sh no longer adds alarms with an invalid date/time.

v0.1 [FB] [Mon Apr 11 09:56:34 CEST 2005]
	* Initial release


alarm.sh
#!/bin/bash
#
# alarm.sh
#
# Small shell script to inform you of events at certain times.
#
# Example uses:
#  $ alarm.sh -n -t "20:15" -m "Meet and greet with John in 15 minutes" 
#  $ alarm.sh -n -r -t "friday" -m "Friday beer-drinking competition tonight"
#  $ alarm.sh -c -i2 -t "thursday" -m "Thursday beer-drinking competition tonight"
#
# Start the script with --daemon under X to start the alarm service that will 
# warn you
#
#-----------------------------------------------------------------------------
#
# Copyright (C), 2005 Ferry Boender. Released under the General Public License
# For more information, see the COPYING file supplied with this program.                                                          
#
# Todo:
# - Error checking
#   - Input dates
#   - Availability of tools (xmessage, dialog, gdialog)
#   - Availability of X
# - Some recurring dates are not handled yet: "1 may 20:00" fails
# - GTK / dialog frontend.
# - Daemon kill mode
# - More
# 

DEFAULT_A_FILE="log.txt"
DEFAULT_ACT="help"

function help() {
	echo "alarm.sh - Alert service client and daemon                     "
	echo "                                                               "
	echo "Usage: alarm.sh [ACTION] [PARAMETERS]                          "
	echo "                                                               "
	echo "Actions:                                                       "
	echo "  -n, --new    (-t, -m)       Create a new alert at timespec   "
	echo "  -c, --change (-i, -t, -m)   Change the alert identified by id"
	echo "  -d, --delete (-i)           Delete the alert identified by id"
	echo "  -l, --list   ([-i])         List alert(s)                    "
	echo "      --daemon                Start the alert daemon           "
  echo "      --weather               Show weather                     "
	echo "      --sun                   Show sunrise and sunset           "
	echo "                                                               "
	echo "Parameters                                                     "
	echo "  -t, --timespec TIMESPEC     Date/time specification at which "
	echo "                              the alert should be triggered    "
	echo "  -m, --message MESSAGE       Message that should be displayed "
	echo "  -i, --id ID                 Alarm ID that should be changed, "
	echo "                              deleted or listed                "
	echo "  -f, --file FILE             Use FILE as alarm file instead of"
	echo "                              the default ~/.alarms            "
	echo "  -r, --repeat                Repetative alarm                 "
}

function error() {
	echo $1
}

function alarmNew() {
	if [ -z "$A_TIMESPEC" ]; then
		error "No timespec specified."
		return
	fi
	if [ -z "$A_MESSAGE" ]; then
		error "No message specified."
		return
	fi
	
	if [ -z "$A_REPEAT" ]; then
		A_TIMESTAMP=`date -d"$A_TIMESPEC" +"%Y-%m-%d %H:%M" 2>/dev/null`
		if [ "$?" -eq "1" ]; then
			echo "Invalid date/time specified";
			exit
		fi
		A_MODE='n'
	else
		A_TIMESTAMP="$A_TIMESPEC";
		A_MODE='r'
	fi

	# Get a new id if need be
	if [ -z "$A_ID" ]; then
		if [ -e "$A_FILE" ]; then
			MAXID=`sort -n "$A_FILE" 2>/dev/null | tail -n1 | cut -d"|" -f1`
			if [ -z "$MAXID" ]; then
				MAXID="0"
			fi
			A_NEWID=`expr $MAXID + 1`
		else
			A_NEWID="1"
		fi
	else
		A_NEWID="$A_ID"
	fi
	
	echo "$A_NEWID|$A_MODE|$A_TIMESTAMP|$A_MESSAGE" >> $A_FILE
	if [ "$SILENT" == "" ]; then
		echo "New alert added (ID $A_NEWID)"
	fi

}

function alarmChange() {
	if [ -z "$A_ID" ]; then
		error "No alarm ID specified."
		return
	fi
	if [ -z "$A_TIMESPEC" ]; then
		error "No timespec specified."
		return
	fi
	if [ -z "$A_MESSAGE" ]; then
		error "No message specified."
		return
	fi
	
	alarmDelete
	alarmNew
}

function alarmDelete() {
	if [ -z "$A_ID" ]; then
		error "No alarm ID specified."
		return
	fi
	
	TMPFILE=`tempfile`
	
	grep -v "^$A_ID|" $A_FILE > $TMPFILE 2>/dev/null 
	mv $TMPFILE $A_FILE

	if [ "$SILENT" == "" ]; then
		echo "Deleted alert (ID $A_ID)"
	fi
}

function alarmList() {
	echo " id | m |                 date | message"
	echo "----+---+----------------------+---------------------------------------------"
	
	cat $A_FILE 2>/dev/null | while read; do
		IFS_OLD=$IFS # Work around for non-working -d param on STUPID Bash read
		IFS="|"
		echo "$REPLY" | { read A_ID A_MODE A_TIMESTAMP A_MESSAGE; # Work around for sub-shell
			if [ "$A_MODE" == "n" ]; then
				A_TIME=`date -d"$A_TIMESTAMP" +"%d %b %Y %H:%M"`
				A_MODE=" "
			else 
				A_TIME="$A_TIMESTAMP"
			fi

			printf "%3s | $A_MODE | %20s | $A_MESSAGE\n" "$A_ID" "$A_TIMESTAMP"
		}
		IFS=$IFS_OLD
	done;
}

function daemon() {
	SILENT=1

	while true; do
		# Current time
		N_TIMESTAMP=`date +"%Y%m%d%H%M"`
		
		cat $A_FILE 2>/dev/null | while read; do
			IFS_OLD=$IFS # Work around for non-working -d param on STUPID Bash read
			IFS="|"
			echo "$REPLY" | { read A_ID A_MODE A_TIMESTAMP A_MESSAGE; # Work around for sub-shell
				A_TIMESTAMP=`date -d"$A_TIMESTAMP" +"%Y%m%d%H%M"`
				
				if [ "$A_MODE" == "n" ]; then
					# Non-repetitive alarm; alert when overdue
					if [ "$N_TIMESTAMP" -ge "$A_TIMESTAMP" ]; then
						xmessage "A $A_MESSAGE"
						alarmDelete "$A_ID"
					fi
				else
					# Repetitive alarm; show only once at exact time
					if [ "$N_TIMESTAMP" -eq "$A_TIMESTAMP" ]; then
						xmessage "A $A_MESSAGE"
					fi
				fi
			}
			IFS=$IFS_OLD
		done;
		sleep 5
	done
}

#here is sunrise-sunset
function alarmSun() {
	#!/bin/bash 
	#if [ -z "$A_ID" ]; then
	#	error "No alarm ID specified."
		#return
	#fi
	
	#TMPFILE=`tempfile`
	
	#grep -v "^$A_ID|" $A_FILE > $TMPFILE 2>/dev/null 
	#mv $TMPFILE $A_FILE

	#if [ "$SILENT" == "" ]; then
	#	echo "Deleted alert (ID $A_ID)"
	#fi
#sunrise-sunset
# Author: Lubos Rendek web@linuxconfig.org

# First obtain a location code from: https://weather.codes/search/

# Insert your location. For example LOXX0001 is a location code for Bratislava, Slovakia
location="KSXX0037"
# Obtain sunrise and sunset raw data from weather.com
sun_times=$( lynx --dump  https://weather.com/weather/today/l/$location | grep "\* Sun" | sed "s/[[:alpha:]]//g;s/*//" )
# Alternative curl/wget solution. In case you wish to use curl or wget instead of lynx comment the above line and uncomment one of the lines below:
#sun_times=$( curl -s  https://weather.com/weather/today/l/$location | sed 's/<span/\n/g' | sed 's/<\/span>/\n/g'  | grep -E "dp0-details-sunrise|dp0-details-sunset" | tr -d '\n' | sed 's/>/ /g' | cut -d " " -f 4,8 )
#sun_times=$( wget -qO-  https://weather.com/weather/today/l/$location | sed 's/<span/\n/g' | sed 's/<\/span>/\n/g'  | grep -E "dp0-details-sunrise|dp0-details-sunset" | tr -d '\n' | sed 's/>/ /g' | cut -d " " -f 4,8 )
# Extract sunrise and sunset times and convert to 24 hour format
	echo "Sunrise and Sunset in Seoul"
sunrise=$(date --date="`echo $sun_times | awk '{ print $1}'` AM" +%R)
sunset=$(date --date="`echo $sun_times | awk '{ print $2}'` PM" +%R)


# Use $sunrise and $sunset variables to fit your needs. Example:
echo "Sunrise for location $location: $sunrise"
echo "Sunset for location $location: $sunset".
}

#here is weather
function alarmWeather() {
#!/bin/bash 

echo "weather"
echo -e "\e[36m###############################################\e[0m"
toilet -f smblock --filter border:metal "WeaTher Forecast"
echo -e "\e[36m###############################################\e[0m"
echo -e "\e[1;31;42m ############ ©Technical Dada ############\e[0m"
echo -e "\e[1;31mUse Landscape For Mobile Devices\e[0m"
echo -e "\e[36m###############################################\e[0m"
echo -e "\e[1;33;4;44m Enter the name of the city you live or near city\e[0m"
read town
echo " ">report
curl wttr.in/$town --connect-timeout 15 -o report -s
count=$(< "report" wc -l)
if [ $count -gt 2 ]
then
echo $'\n\n\n'
head -n -1 report
else
echo -e "\e[1;31mConnection Timed Out..."
echo -e "Seems That Your Input is Wrong or Net is Slow\e[0m"
fi
rm report >/dev/null 2>&1
echo 

}

########################################################################
# Parse commandline options
########################################################################

TEMP=`getopt -o ncdlt:m:i:f:rh --long new,change,delete,list,timespec:,message:,id:,file:,repeat,daemon,help \
     -n 'alarm.sh' -- "$@"`

if [ $? != 0 ] ; then echo "Terminating..." >&2 ; exit 1 ; fi

eval set -- "$TEMP"

while true ; do
	case "$1" in
		-n|--new)      ACT="new";      shift 1 ;;
		-c|--change)   ACT="change";   shift 1 ;;
		-d|--delete)   ACT="delete";   shift 1 ;;
		-l|--list)     ACT="list";     shift 1 ;;
       --sun)      ACT="sun";      shift 1 ;;
       --weather)  ACT="weather";  shift 1 ;;
		   --daemon)   ACT="daemon";   shift 1 ;;
		-t|--timespec) A_TIMESPEC=$2;  shift 2 ;;
		-m|--message)  A_MESSAGE=$2;   shift 2 ;;
		-i|--id)       A_ID=$2;        shift 2 ;;
		-f|--file)     A_FILE=$2;      shift 2 ;;
		-r|--repeat)   A_REPEAT="r";   shift 1 ;;
		-h|--help)     FLAG_HELP=1;    shift 1 ;;
		--)                            shift 1; break ;;
		*)             echo "Internal error!" ; exit 1 ;;
	esac
done

# Default action
if [ -z "$ACT" ]; then
	ACT="$DEFAULT_ACT"
fi
# Default alarm file
if [ -z "$A_FILE" ]; then
	A_FILE="$DEFAULT_A_FILE"
fi

case "$ACT" in
	new)     alarmNew ;;
	change)  alarmChange ;;
	delete)  alarmDelete ;;
	list)    alarmList ;;
  sun)     alarmSun ;;
  weather) alarmWeather ;;
	help)    help ;;
	daemon)  daemon ;;
esac

exit

A_FILE=$1

ALARM=`date -d"+1 minute" +"%Y%m%d%H%M"`

while true; do
	# Current time
	N_TIMESTAMP=`date +"%Y%m%d%H%M"`
	
	# Alert times
	cat $A_FILE | while read; do
		A_TIME=`echo "$REPLY" | cut -d"|" -f1`
		A_MSG=`echo "$REPLY" | cut -d"|" -f2-`
		A_TIMESTAMP=`date -d"$A_TIME" +"%Y%m%d%H%M"`

		# Trigger any alerts if need be
		if [ "$N_TIMESTAMP" -ge "$A_TIMESTAMP" ]; then
			A_NICETIME=`date -d"$A_TIME" +"%d %m %Y %H:%M"`
			TMPFILE=`tempfile`

			# Remove entry from file
			grep -v "$REPLY" $A_FILE > $TMPFILE
			mv $TMPFILE $A_FILE

			# Trigger alert
			xmessage "$A_NICETIME : $A_MSG"
		fi
	done;
	sleep 30
done
